File: LAB6/Expression.cpp
#include "Expression.h"
#include <iostream>
#include <math.h>

using namespace std;

Expression::Expression(double a,  double c, double d)
{
	this->a = a;
	this->c = c;
	this->d = d;
}

double Expression::Calculate()
{
	if ((2 * c - a) < 0)
	{
		throw domain_error("Logarithm from negative value not possible");
	}
	if ((a / 4 + c) == 0)
	{
		throw "Division by zero exception";
	}

	return log((2 * c - a)  + d - 152);
}

Expression* Expression::operator +(int num)
{
	return new Expression(this->a + num,  this->c + num, this->d + num);
}

Expression* Expression::operator +(Expression example)
{
	return new Expression(this->a + example.a,  this->c + example.c, this->d + example.d);
}

Expression* Expression::operator -(int num)
{
	return new Expression(this->a - num,  this->c - num, this->d - num);
}

Expression* Expression::operator -(Expression example)
{
	return new Expression(this->a - example.a,  this->c - example.c, this->d - example.d);
}

File: LAB6/Expression.cs
using System;

namespace LAB6_OOP_CS
{
    class Expression
    {
       

        public double A, C, D;
        
        public Expression(double a,  double c, double d)
        {
            A = a;
            C = c;
            D = d;
        }

        public double Сalculating()
        {
            
            if  ((2 * C - A) < 0)
            {
                throw new ArgumentOutOfRangeException();
            }
            if ((A / 4 + C) == 0)
            {
                throw new DivideByZeroException();
            }

            
            return (Math.Log((2 * C - A), 10) + D - 152) / (A / 4 + C);

        }
        
        #region
        public static Expression operator +(Expression example, int num)
        {
            return new Expression(example.A + num, example.C + num, example.D + num);
        }
        public static Expression operator +(Expression example1, Expression example2)
        {
            return new Expression(example1.A + example2.A, example1.C + example2.C, example1.D + example2.D);
        }
        public static Expression operator -(Expression example, int num)
        {
            return new Expression(example.A - num, example.C - num, example.D - num);
        }
        public static Expression operator -(Expression example1, Expression example2)
        {
            return new Expression(example1.A - example2.A, example1.C - example2.C, example1.D - example2.D);
        }
        #endregion

    }
}